### PR TITLE
[extensions/storage/filestorage]: add compaction on start

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**
Add compaction to file storage extension.

**Link to tracking Issue:**

#7766 

**Testing:**
 - manual done
 - ut In progress

**Documentation:**
N/A

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+- `storageextension`: add optional compaction to filestorage (#7768)
```
Comment (pmm-sumo) [inline CHANGELOG.md:5]: Should be under Enhancements I believe :)

File: extension/storage/filestorage/config.go [hunk 2]
```diff
@@ -29,6 +29,9 @@ type Config struct {
 
 	Directory string        `mapstructure:"directory,omitempty"`
 	Timeout   time.Duration `mapstructure:"timeout,omitempty"`
+
+	CompactOnStart      bool   `mapstructure:"compact_on_start,omitempty"`
+	CompactionDirectory string `mapstructure:"compaction_directory,omitempty"`
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply extension/storage/filestorage/config.go:34]: yup, added to factory

**Comment by pmm-sumo** [comment]
@djaglowski WDYT?

**Review by djaglowski — COMMENTED** [review]
This all makes sense to me. I have some questions on the config and what it should support (now and eventually).

My first thought is that although compacting on start is a great standalone improvement, it would seem that long-running collectors will eventually need this functionality on-the-fly as well. I'm not suggesting that needs to be implemented yet, just stating that we'll possibly need at least one additional config setting (later).

I also mentioned in my other comment that we probably need a setting for the transaction size.

If we're going to end up with 3 or 4 compaction-related settings, I think we should consider organizing them under a single struct. Maybe something like: 
```yaml
file_storage:
  directory: /var/lib/otelcol/mydir
  compaction:                  # default nil
    directory: ...             # default same directory as db
    on_start: true             # default false
    on_interval: 24h           # default disabled (and not yet implemented)
    max_transaction_size: 1000 # default ?
```

The only other thing I'd note is that we really should have a test or two to validate that the old file ends up downsized, and the temp file is deleted.

**Review by djaglowski — APPROVED** [review]
LGTM

**Review by pmm-sumo — CHANGES_REQUESTED** [review]
We should update README.md of the extension. Otherwise, lgtm

**Comment by sumo-drosiek** [comment]
@pmm-sumo please look again

File: extension/storage/filestorage/config.go [hunk 3]
```diff
@@ -29,24 +30,44 @@ type Config struct {
...
-			)
+		if !info.IsDir() {
+			return fmt.Errorf("%s is not a directory", dir)
 		}
-
 	}
-	if !info.IsDir() {
-		return fmt.Errorf("%s is not a directory", cfg.Directory)
+
+	if cfg.Compaction.MaxTransactionSize < 0 {
```
Comment (pmm-sumo) [inline extension/storage/filestorage/config.go:69]: Can we use default value? Is 0 a valid value?

Comment (sumo-drosiek) [reply extension/storage/filestorage/config.go:69]: A value of zero will ignore transaction sizes
https://github.com/etcd-io/bbolt/blob/master/compact.go

Comment (pmm-sumo) [reply extension/storage/filestorage/config.go:69]: I had in mind a default non-zero transaction size (if it was not set). But I believe zero should be fine as default

**Review by pmm-sumo — APPROVED** [review]
Changelog conflict needs rebase

**Comment by sumo-drosiek** [comment]
> Changelog conflict needs rebase
rebased

**Pull request merged**